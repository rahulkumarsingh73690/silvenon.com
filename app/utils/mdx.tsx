import { useMemo } from 'react'
import { getMDXComponent } from 'mdx-bundler/client'
import Gitgraph from '../components/Gitgraph'
import ProseImage from '../components/ProseImage'
import HotTip from '../components/HotTip'
import { Tweet } from 'react-twitter-widgets'
import { MDXSpan, MDXPre } from '../components/pretty-code'
import { useDarkMode } from '~/services/dark-mode'

export function useMDXPost(code: string) {
  const Post = useMemo(() => getMDXComponent(code), [code])
  return (
    <Post
      components={{
        Gitgraph,
        ProseImage,
        HotTip,
        Tweet: ({ tweetId }: { tweetId: string }) => {
          const darkMode = useDarkMode()
          return (
            <Tweet
              tweetId={tweetId}
              options={darkMode ? { theme: 'dark' } : {}}
            />
          )
        },
        // @ts-expect-error
        span: MDXSpan,
        // @ts-expect-error
        pre: MDXPre,
      }}
    />
  )
}
