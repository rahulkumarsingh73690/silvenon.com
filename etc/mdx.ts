import customBlocks from 'remark-custom-blocks'
import smartypants from '@silvenon/remark-smartypants'
import unwrapImages from 'remark-unwrap-images'
import prism from 'mdx-prism'

export const mdxOptions = {
  remarkPlugins: [
    [
      customBlocks,
      {
        tip: {
          classes: 'tip',
          title: 'required',
        },
      },
    ],
    smartypants,
    unwrapImages,
  ],
  rehypePlugins: [prism],
}
