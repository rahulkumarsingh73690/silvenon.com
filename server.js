const express = require('express')
const next = require('next')

const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev })
const handle = app.getRequestHandler()

const redirects = [
  { from: '/intro-to-eslint', to: 'https://blog.silvenon.com/intro-to-eslint-81f6c9f24ce5' },
  { from: '/testing-react-and-redux', to: 'https://semaphoreci.com/community/series/how-to-test-a-react-and-redux-application-with-ava' },
  { from: '/testing-react-and-redux-pt1', to: 'https://semaphoreci.com/community/tutorials/getting-started-with-create-react-app-and-ava' },
  { from: '/testing-react-and-redux-pt2', to: 'https://semaphoreci.com/community/tutorials/testing-common-redux-patterns-in-react-using-ava' },
  { from: '/testing-react-and-redux-pt3', to: 'https://semaphoreci.com/community/tutorials/testing-react-components-with-ava' },
]

app.prepare().then(() => {
  const server = express()

  redirects.forEach(({ from, to }) => {
    server.get(from, (req, res) => {
      res.redirect(301, to)
    })
  })

  server.get('*', (req, res) => {
    handle(req, res)
  })

  server.listen(3000, (err) => {
    if (err) throw err
    // eslint-disable-next-line no-console
    console.log('> Ready on http://localhost:3000')
  })
})
