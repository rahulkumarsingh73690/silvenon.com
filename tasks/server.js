const browserSync = require('browser-sync')
const { destDir } = require('../site.config')

const server = browserSync.create()
const port = 3000
const path = '/'

function initServer(done) {
  server.init(
    {
      port,
      server: destDir,
      ui: false,
      open: false,
      notify: false,
      files: [
        `${destDir}/**/*.html`,
        `${destDir}/**/*.css`,
        `${destDir}/**/*.js`,
      ],
    },
    done,
  )
}

function reloadServer(done) {
  server.reload()
  done()
}

function exitServer(done) {
  server.exit()
  done()
}

module.exports = {
  init: initServer,
  exit: exitServer,

  stream: server.stream,
  reload: reloadServer,
  port,
  path,
}
