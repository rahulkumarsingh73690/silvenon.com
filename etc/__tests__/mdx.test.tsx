import React from 'react'
import compileMdx from '@mdx-js/mdx'
import { mdx } from '@mdx-js/react'
import { renderToString } from 'react-dom/server'
import dedent from 'dedent'
import { transform as babelTransform } from '@babel/core'
import { mdxOptions } from '../mdx'

// https://github.com/mdx-js/mdx/blob/d140f95eca338a4f4722b095278f021c6c248f45/packages/react/test/test.js

async function getMdxComponent(mdxContent: string, options: any) {
  // Turn the serialized MDX code into serialized JSX…
  const jsxContent = await compileMdx(mdxContent, {
    skipExport: true,
    ...options,
  })

  // …and that into serialized JS.
  const result = babelTransform(jsxContent, {
    configFile: false,
    plugins: [
      '@babel/plugin-transform-react-jsx',
      'babel-plugin-remove-export-keywords',
    ],
  })

  if (result) {
    // …and finally run it, returning the component.
    return new Function('mdx', `${result.code}; return MDXContent`)(mdx)
  }
}

describe('MDX processor', () => {
  test('output', async () => {
    const MDXComponent = await getMdxComponent(
      dedent`
        Introduction to my post with smart -- punctuation.

        ## Section

        ~~~js{2}
        function doSomething() {
          console.log('hello')
        }
        ~~~

        [[tip | Hot tip!]]
        | Content.
        |
        | More content.
      `,
      mdxOptions,
    )

    const template = document.createElement('template')
    template.innerHTML = renderToString(<MDXComponent />)
    expect(template.content).toMatchSnapshot()
  })
})
