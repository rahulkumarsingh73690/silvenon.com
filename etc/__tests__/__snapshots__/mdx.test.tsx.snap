// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDX processor output 1`] = `
<DocumentFragment>
  <p>
    Introduction to my post with smart â€” punctuation.
  </p>
  <h2>
    Section
  </h2>
  <pre
    class="language-js"
  >
    <code
      class="language-js{2}"
    >
      <span
        class="token keyword"
      >
        function
      </span>
       
      <span
        class="token function"
      >
        doSomething
      </span>
      <span
        class="token punctuation"
      >
        (
      </span>
      <span
        class="token punctuation"
      >
        )
      </span>
       
      <span
        class="token punctuation"
      >
        {
      </span>
      

      <div
        class="mdx-marker"
      >
          
        <span
          class="token console class-name"
        >
          console
        </span>
        <span
          class="token punctuation"
        >
          .
        </span>
        <span
          class="token method function property-access"
        >
          log
        </span>
        <span
          class="token punctuation"
        >
          (
        </span>
        <span
          class="token string"
        >
          'hello'
        </span>
        <span
          class="token punctuation"
        >
          )
        </span>
        

      </div>
      <span
        class="token punctuation"
      >
        }
      </span>
      

    </code>
  </pre>
  <div
    class="custom-block tip"
  >
    <div
      class="custom-block-heading"
    >
      Hot tip!
    </div>
    <div
      class="custom-block-body"
    >
      <p>
        Content.
      </p>
      <p>
        More content.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
